lib/ansible/plugins/filter/core.py

```diff
@@ -308,19 +331,21 @@ def to_uuid(string, namespace=UUID_NAMESPACE_ANSIBLE):
     return to_text(uuid.uuid5(uuid_namespace, to_native(string, errors='surrogate_or_strict')))
 
 
+@accept_marker
 def mandatory(a, msg=None):
     """Make a variable mandatory."""
-    from jinja2.runtime import Undefined
+    # DTFIX-MERGE: deprecate this filter; there are much better ways via undef, etc...
+    #              also remember to remove unit test checking for _undefined_name
+    if isinstance(a, UndefinedMarker):
+        if msg is not None:
+            raise AnsibleFilterError(to_text(msg))
 
-    if isinstance(a, Undefined):
         if a._undefined_name is not None:
-            name = "'%s' " % to_text(a._undefined_name)
+            name = f'{to_text(a._undefined_name)!r} '
         else:
             name = ''
 
-        if msg is not None:
-            raise AnsibleFilterError(to_native(msg))
-        raise AnsibleFilterError("Mandatory variable %s not defined." % name)
+        raise AnsibleFilterError(f"Mandatory variable {name}not defined.")
```

Comment:
```suggestion
        raise AnsibleFilterError(f"Mandatory variable {name} not defined.")
```

Reply:
`name` is either empty or has a space already appended.

Reply:
The `name` already contains a trailing space, or is empty, so no additional space is needed.